lib/libc/minimal/source/stdlib/strtol.c

```diff
@@ -98,9 +98,9 @@ long strtol(const char *nptr, char **endptr, register int base)
 	cutoff /= (unsigned long)base;
 	for (acc = 0, any = 0;; c = *s++) {
 		if (isdigit((unsigned char)c)) {
-			c -= '0';
+			c = (char)c - '0';
```

Comment:
Can you explain why changing the type of 'c' isn't the appropriate change here and in the other locations in this file? Casting like this might be silencing the checker, but it makes the code much harder to follow.

Reply:
To subtract a character from an integer makes little sense, while to subtract a character from a character is natural way to compute the distance between two character codes. This is the MISRA perspective and I found it is very sensible.

Reply:
That's not what I asked. Why isn't the correct fix here to change the type of `c` to `unsigned char`?

Reply:
In MISRA perspective <code>unsigned char</code> is a small integer type and not an essentially character type, so it would have no effect for MISRA violation on line 101 (it would be still a subtraction between an integer and a character).

Reply:
That's ridiculous. You need it to be unsigned char for the isdigit call, and a signed char for the calculation?!!?

Reply:
I suggest to keep the conversation on a technical level and avoid epithets.

1) the reason for unsigned char cast is in C standard:
C18 7.4p1:
The header <ctype.h> declares several functions useful for classifying and mapping characters.
In all cases the argument is an int , the value of which shall be representable as an unsigned char
or shall equal the value of the macro EOF . If the argument has any other value, the behavior is
undefined.

2) the reason for the use of plain char (not signed char) for character difference is in MISRA:
Amplification of Rule 10.2:
For  the  -  operator,  the  first  operand  shall  have  essentially  character  type  and  the  second  
shall  have  essentially  signed  type,  essentially  unsigned  type  or  essentially  character  type.  If  both  
operands have essentially character type then the result has the standard type (usually int in this 
case) else the result has essentially character type.

As I've explained in note above the rationale is that if character are seen as ordered symbols, to compute the integer offset between two symbols we have to do char - char.

Reply:
So, with the tool I am able to run here (cppcheck + misra.py), I don't get any report of violation of 10.2 (nor 10.4) if I change the type of `c`:

```
diff --git a/lib/libc/minimal/source/stdlib/strtoul.c b/lib/libc/minimal/source/stdlib/strtoul.c
index 0458ca94b5..4bd1380978 100644
--- a/lib/libc/minimal/source/stdlib/strtoul.c
+++ b/lib/libc/minimal/source/stdlib/strtoul.c
@@ -47,7 +47,7 @@ unsigned long strtoul(const char *nptr, char **endptr, register int base)
 {
        register const char *s = nptr;
        register unsigned long acc;
-       register int c;
+       register unsigned char c;
        register unsigned long cutoff;
        register int neg = 0, any, cutlim;
```

Maybe your tooling is behaving differently or is pickier? cppcheck seems to be acknowledged as an official misra compliance test these days (URL moved so archive.org link is here):

https://web.archive.org/web/20210416015646/https://www.misra.org.uk/Links/tabid/63/Default.aspx

Is there a way to run it to see output in CI of the licensed tools to compare? Thanks!

Reply:
Any tool that is not flagging a MISRA C:2012 Rule 10.4 violation for, e.g., (c == '0') when c is declared of type unsigned char is a tool with false negatives for that rule.

Reply:
Sure, but it flags 10.4 even when `c` is an int as of today, doesn't it?

These are all circular restrictions that in the end either means casts everywhere (and the code readability impact of that), or something like an inline helper or macro that casts for you. The latter seems like less churn.

Reply:
There will not be casts everywhere. It was agreed to proceed one rule at a time: let's talk again about that when it will be the turn of Rule 10.4 (no, the changes for Rule 10.4 will not introduce any new violations for the rules that have already been dealt with).